io-engine/src/lvs/lvs_lvol.rs

```diff
@@ -443,15 +442,108 @@ impl Lvol {
+    pub fn list_all_snapshots() -> Vec<VolumeSnapshotDescriptor> {
+        let mut snapshot_list: Vec<VolumeSnapshotDescriptor> = Vec::new();
+        let bdev =
+            UntypedBdev::bdev_first().expect("Failed to enumerate devices");
+
+        let lvol_devices = bdev
+            .into_iter()
+            .filter(|b| b.driver() == "lvol")
+            .map(|b| Lvol::try_from(b).unwrap())
+            .collect::<Vec<Lvol>>();
+        // return empty snapshot parameter list, if no snapshot found.
+        if lvol_devices.len() <= 1 {
+            return snapshot_list;
+        }
+        for lvol in lvol_devices {
+            // skip lvol if it is not snapshot.
+            if !lvol.is_snapshot() {
+                continue;
+            }
+            let blob = lvol.bs_iter_first();
+            let mut skip_lvol = false;
+            let mut snapshot_param: SnapshotParams = Default::default();
+            for attr in SnapshotXattrs::iter() {
+                let mut val = std::ptr::null();
+                let mut size: u64 = 0;
+                let attr_id = attr.name().to_string().into_cstring();
+                let curr_attr_val;
+                unsafe {
+                    let r = spdk_blob_get_xattr_value(
+                        blob,
+                        attr_id.as_ptr(),
+                        &mut val as *mut *const c_void,
+                        &mut size as *mut u64,
+                    );
+                    // skip current lvol if any of the snapshot attribute not
+                    // found.
+                    if r != 0 {
+                        skip_lvol = true;
+                        warn!(
+                            "Snapshot attribute {:?} not found, skip Lvol{:?}",
+                            attr_id, lvol
+                        );
+                        break;
+                    }
+                    curr_attr_val = String::from_raw_parts(
+                        val as *mut u8,
+                        size as usize,
+                        size as usize,
+                    );
+                }
+                match attr {
+                    SnapshotXattrs::ParentId => {
+                        snapshot_param.set_parent_id(curr_attr_val);
+                    }
+                    SnapshotXattrs::EntityId => {
+                        snapshot_param.set_entity_id(curr_attr_val);
+                    }
+                    SnapshotXattrs::TxId => {
+                        snapshot_param.set_txn_id(curr_attr_val);
+                    }
+                }
+            }
+            // skip current lvol if any of the snapshot attribute not found.
+            if skip_lvol {
+                continue;
+            }
+            // set remaining snapshot parameters for snapshot list
+            snapshot_param.set_name(lvol.name());
+            let replica_lvol = Bdev::lookup_by_uuid_str(
+                snapshot_param.parent_id().unwrap().as_str(),
```

Comment:
should this unwrap be handled?

Reply:
As this is the list... should we somehow return "broken" snapshots as well, with just uuid set so at least CP is aware of these "broken" snapshots?

Reply:
Added a new flag to track the broken snapshots and return the flag to control plane. Please check once.